addons/point_of_sale/models/pos_session.py

```diff
@@ -305,22 +306,26 @@ def _validate_session(self):
             self._check_if_no_draft_orders()
             if self.update_stock_at_closing:
                 self._create_picking_at_end_of_session()
-            # Users without any accounting rights won't be able to create the journal entry. If this
-            # case, switch to sudo for creation and posting.
-            sudo = False
-            if (
-                not self.env['account.move'].check_access_rights('create', raise_exception=False)
-                and self.user_has_groups('point_of_sale.group_pos_user')
-            ):
-                sudo = True
-                self.sudo().with_company(self.company_id)._create_account_move()
-            else:
-                self.with_company(self.company_id)._create_account_move()
-            if self.move_id.line_ids:
-                # Set the uninvoiced orders' state to 'done'
-                self.env['pos.order'].search([('session_id', '=', self.id), ('state', '=', 'paid')]).write({'state': 'done'})
-            else:
-                self.move_id.unlink()
+            try:
+                with self.env.cr.savepoint():
+                    if (
+                        not self.env['account.move'].check_access_rights('create', raise_exception=False)
+                        and self.user_has_groups('point_of_sale.group_pos_user')
+                    ):
+                        self.sudo().with_company(self.company_id)._create_account_move(balancing_account, amount_to_balance)
+                    else:
+                        self.with_company(self.company_id)._create_account_move(balancing_account, amount_to_balance)
+                    if self.move_id.line_ids:
+                        # Set the uninvoiced orders' state to 'done'
+                        self.env['pos.order'].search([('session_id', '=', self.id), ('state', '=', 'paid')]).write({'state': 'done'})
+                    else:
+                        self.move_id.unlink()
+
+            except ImbalanceJournalEntryError as error:
+                if len(error.imbalance_amounts) != 1:
+                    raise UserError(_("Single imbalance amount that matches the pos session's account move is expected."))
```

Comment:
This seems like a programming error that won't be of any use to the customer.
I like to raise a general traceback in those cases, making it easier to debug; no need to translate it either. You will be able to find the problem without having to search for it even with just a screenshot of a client speaking a language you don't understand :wink:

Reply:
@caburj keep translation of usererror